File: WayFinder/ViewModels/ReelsViewModel.swift
import Foundation
import SwiftUI

@MainActor
class ReelsViewModel: ObservableObject {
    @Published var uiState: ReelsUiState = .idle
    @Published var items: [ReelContentItem] = []
    
    private var currentPage = 0
    private let pageSize = 10
    private var hasMore = true
    
    private let discussionService = DiscussionService.shared
    private let journeyService = JourneyService.shared
    @Published private(set) var currentUserId: String?
    
    init() {
        loadCurrentUserId()
    }
    
    private func loadCurrentUserId() {
        Task {
            do {
                let profile = try await UserService.shared.fetchProfile()
                await MainActor.run {
                    self.currentUserId = profile.id
                }
            } catch {
                print("⚠️ [ReelsViewModel] Could not load current user ID: \(error)")
            }
        }
    }
    
    func loadReelsFeed(refresh: Bool = false) async {
        do {
            if refresh {
                currentPage = 0
                hasMore = true
            }
            
            if !hasMore && !refresh {
                return
            }
            
            uiState = .loading
            
            // Fetch posts and journeys in parallel
            async let postsTask = fetchPosts()
            async let journeysTask = fetchJourneys()
            
            let (postsResult, journeysResult) = try await (postsTask, journeysTask)
            
            // Convert to ReelContentItem
            let postItems = postsResult.posts.map { post in
                // Debug: Print post user info
                let user = post.user
                print("📸 [ReelsViewModel] Post \(post.id) - User: \(user.id), Username: \(user.username ?? "nil"), Avatar: \(user.profileImageUrl ?? "nil")")
                return ReelContentItem.postItem(post)
            }
            let journeyItems = journeysResult.map { journey in
                // Debug: Print journey user info
                if let user = journey.user {
                    print("📸 [ReelsViewModel] Journey \(journey.id) - User: \(user.id), Username: \(user.username), Avatar: \(user.profileImageUrl ?? "nil")")
                } else {
                    print("⚠️ [ReelsViewModel] Journey \(journey.id) - User is nil")
                }
                return ReelContentItem.journeyItem(journey)
            }
            
            // Merge and sort by engagement (likes + comments) and recency
            let allItems = (postItems + journeyItems).sorted { lhs, rhs in
                let lhsEngagement = lhs.likesCount + lhs.commentsCount
                let rhsEngagement = rhs.likesCount + rhs.commentsCount
                
                if lhsEngagement != rhsEngagement {
                    return lhsEngagement > rhsEngagement
                }
                
                // Sort by recency (newer first)
                return lhs.createdAt > rhs.createdAt
            }
            
            let existingItems: [ReelContentItem] = (refresh || currentPage == 0) ? [] : items
            
            items = existingItems + allItems
            hasMore = allItems.count == pageSize && (postsResult.posts.count == pageSize || journeysResult.count == pageSize)
            
            uiState = .success(items: items, hasMore: hasMore)
            currentPage += 1
        } catch {
            print("❌ [ReelsViewModel] Error loading reels feed: \(error)")
            uiState = .error(message: error.localizedDescription)
        }
    }
    
    private func fetchPosts() async throws -> (posts: [DiscussionPost], hasMore: Bool) {
        do {
            let posts = try await discussionService.getPosts(
                limit: pageSize,
                skip: currentPage * pageSize,
                destination: nil
            )
            return (posts, posts.count == pageSize)
        } catch {
            print("❌ [ReelsViewModel] Error loading posts: \(error)")
            return ([], false)
        }
    }
    
    private func fetchJourneys() async throws -> [Journey] {
        do {
            return try await journeyService.getJourneys(
                limit: pageSize,
                skip: currentPage * pageSize
            )
        } catch {
            print("❌ [ReelsViewModel] Error loading journeys: \(error)")
            return []
        }
    }
    
    func likeItem(_ item: ReelContentItem) async {
        do {
            switch item {
            case .postItem(let post):
                let updatedPost = try await discussionService.likePost(id: post.id)
                // Update state optimistically
                if let index = items.firstIndex(where: { $0.id == item.id }) {
                    if case .postItem = items[index] {
                        items[index] = .postItem(updatedPost)
                        uiState = .success(items: items, hasMore: hasMore)
                    }
                }
            case .journeyItem(let journey):
                _ = try await journeyService.likeJourney(journeyId: journey.id)
                // Reload journey to get updated state
                let updatedJourney = try await journeyService.getJourney(by: journey.id)
                // Update state optimistically
                if let index = items.firstIndex(where: { $0.id == item.id }) {
                    if case .journeyItem = items[index] {
                        items[index] = .journeyItem(updatedJourney)
                        uiState = .success(items: items, hasMore: hasMore)
                    }
                }
            }
        } catch {
            print("❌ [ReelsViewModel] Error liking item: \(error)")
            // Reload on error
            await loadReelsFeed(refresh: true)
        }
    }
    
    func retry() {
        Task {
            await loadReelsFeed(refresh: true)
        }
    }
}

enum ReelsUiState {
    case idle
    case loading
    case success(items: [ReelContentItem], hasMore: Bool)
    case error(message: String)
}

File: app/src/main/java/tn/esprit/wayfinder/viewmodels/ReelsViewModel.kt
package tn.esprit.wayfinder.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import tn.esprit.wayfinder.models.DiscussionPost
import tn.esprit.wayfinder.models.Journey
import tn.esprit.wayfinder.presentation.discussion.DiscussionRepository
import tn.esprit.wayfinder.presentation.journey.JourneyRepository

/**
 * Unified content item for reels feed
 */
sealed class ReelContentItem {
    data class PostItem(val post: DiscussionPost) : ReelContentItem()
    data class JourneyItem(val journey: Journey) : ReelContentItem()
    
    val id: String
        get() = when (this) {
            is PostItem -> "post_${post.id}"
            is JourneyItem -> "journey_${journey.id}"
        }
    
    val likesCount: Int
        get() = when (this) {
            is PostItem -> post.likesCount
            is JourneyItem -> journey.likesCount
        }
    
    val commentsCount: Int
        get() = when (this) {
            is PostItem -> post.commentsCount
            is JourneyItem -> journey.commentsCount
        }
    
    val createdAt: String
        get() = when (this) {
            is PostItem -> post.createdAt
            is JourneyItem -> journey.createdAt ?: ""
        }
    
    val isLiked: Boolean
        get() = when (this) {
            is PostItem -> false // Will be determined by likedBy list
            is JourneyItem -> journey.isLiked
        }
}

sealed class ReelsUiState {
    object Idle : ReelsUiState()
    object Loading : ReelsUiState()
    data class Success(
        val items: List<ReelContentItem>,
        val hasMore: Boolean = false
    ) : ReelsUiState()
    data class Error(val message: String) : ReelsUiState()
}

class ReelsViewModel(
    private val discussionRepository: DiscussionRepository,
    private val journeyRepository: JourneyRepository
) : ViewModel() {
    
    private val _uiState = MutableStateFlow<ReelsUiState>(ReelsUiState.Idle)
    val uiState: StateFlow<ReelsUiState> = _uiState.asStateFlow()
    
    private var currentPage = 0
    private val pageSize = 10
    private var hasMore = true
    
    fun loadReelsFeed(refresh: Boolean = false) {
        viewModelScope.launch {
            try {
                if (refresh) {
                    currentPage = 0
                    hasMore = true
                }
                
                if (!hasMore && !refresh) {
                    return@launch
                }
                
                _uiState.value = ReelsUiState.Loading
                
                // Fetch posts and journeys in parallel
                val postsDeferred = async {
                    try {
                        discussionRepository.getPosts(
                            limit = pageSize,
                            skip = currentPage * pageSize,
                            destination = null
                        )
                    } catch (e: Exception) {
                        android.util.Log.e("ReelsViewModel", "Error loading posts: ${e.message}", e)
                        null
                    }
                }
                
                val journeysDeferred = async {
                    try {
                        journeyRepository.getJourneys(limit = pageSize, skip = currentPage * pageSize)
                    } catch (e: Exception) {
                        android.util.Log.e("ReelsViewModel", "Error loading journeys: ${e.message}", e)
                        emptyList()
                    }
                }
                
                val postsResponse = postsDeferred.await()
                val journeys = journeysDeferred.await()
                
                // Convert to ReelContentItem
                val postItems = postsResponse?.data?.map { ReelContentItem.PostItem(it) } ?: emptyList()
                val journeyItems = journeys.map { ReelContentItem.JourneyItem(it) }
                
                // Merge and sort by engagement (likes + comments) and recency
                val allItems = (postItems + journeyItems).sortedWith(
                    compareByDescending<ReelContentItem> { 
                        it.likesCount + it.commentsCount 
                    }.thenByDescending { 
                        it.createdAt 
                    }
                )
                
                val existingItems = if (refresh || currentPage == 0) {
                    emptyList()
                } else {
                    (_uiState.value as? ReelsUiState.Success)?.items ?: emptyList()
                }
                
                val updatedItems = existingItems + allItems
                hasMore = allItems.isNotEmpty() && (postsResponse?.data?.size == pageSize || journeys.size == pageSize)
                
                _uiState.value = ReelsUiState.Success(
                    items = updatedItems,
                    hasMore = hasMore
                )
                
                currentPage++
            } catch (e: Exception) {
                android.util.Log.e("ReelsViewModel", "Error loading reels feed: ${e.message}", e)
                _uiState.value = ReelsUiState.Error(
                    e.message ?: "Failed to load reels feed"
                )
            }
        }
    }
    
    fun likeItem(item: ReelContentItem, currentUserId: String?) {
        viewModelScope.launch {
            try {
                when (item) {
                    is ReelContentItem.PostItem -> {
                        discussionRepository.likePost(item.post.id)
                        // Update state optimistically
                        val currentState = _uiState.value
                        if (currentState is ReelsUiState.Success) {
                            val updatedItems = currentState.items.map { currentItem ->
                                if (currentItem.id == item.id && currentItem is ReelContentItem.PostItem) {
                                    val isLiked = currentUserId != null && currentItem.post.likedBy.contains(currentUserId)
                                    ReelContentItem.PostItem(
                                        post = currentItem.post.copy(
                                            likesCount = if (isLiked) currentItem.post.likesCount - 1 else currentItem.post.likesCount + 1,
                                            likedBy = if (isLiked) {
                                                currentItem.post.likedBy.filter { it != currentUserId }
                                            } else {
                                                (currentItem.post.likedBy + (currentUserId ?: "")).distinct()
                                            }
                                        )
                                    )
                                } else {
                                    currentItem
                                }
                            }
                            _uiState.value = currentState.copy(items = updatedItems)
                        }
                    }
                    is ReelContentItem.JourneyItem -> {
                        journeyRepository.likeJourney(item.journey.id)
                        // Update state optimistically
                        val currentState = _uiState.value
                        if (currentState is ReelsUiState.Success) {
                            val updatedItems = currentState.items.map { currentItem ->
                                if (currentItem.id == item.id && currentItem is ReelContentItem.JourneyItem) {
                                    ReelContentItem.JourneyItem(
                                        journey = currentItem.journey.copy(
                                            likesCount = if (currentItem.journey.isLiked) currentItem.journey.likesCount - 1 else currentItem.journey.likesCount + 1,
                                            isLiked = !currentItem.journey.isLiked
                                        )
                                    )
                                } else {
                                    currentItem
                                }
                            }
                            _uiState.value = currentState.copy(items = updatedItems)
                        }
                    }
                }
            } catch (e: Exception) {
                android.util.Log.e("ReelsViewModel", "Error liking item: ${e.message}", e)
                // Reload on error
                loadReelsFeed(refresh = true)
            }
        }
    }
    
    fun incrementCommentCount(itemId: String) {
        val currentState = _uiState.value
        if (currentState is ReelsUiState.Success) {
            val updatedItems = currentState.items.map { currentItem ->
                if (currentItem.id == itemId) {
                    when (currentItem) {
                        is ReelContentItem.PostItem -> {
                            ReelContentItem.PostItem(
                                post = currentItem.post.copy(
                                    commentsCount = currentItem.post.commentsCount + 1
                                )
                            )
                        }
                        is ReelContentItem.JourneyItem -> {
                            ReelContentItem.JourneyItem(
                                journey = currentItem.journey.copy(
                                    commentsCount = currentItem.journey.commentsCount + 1
                                )
                            )
                        }
                    }
                } else {
                    currentItem
                }
            }
            _uiState.value = currentState.copy(items = updatedItems)
        }
    }
    
    fun retry() {
        loadReelsFeed(refresh = true)
    }
}

